### PR TITLE
[Utils] Extract CollectDebugInfoForCloning from CloneFunctionInto

### PR DESCRIPTION
**Pull request by artempyanykh**
Summary:
Consolidate the logic in a single function. We do an extra pass over Instructions but this is necessary to untangle things and extract metadata cloning in a future diff.

Test Plan:
```
$ ninja check-llvm-unit check-llvm
[211/213] Running the LLVM regression tests

Testing Time: 106.06s

Total Discovered Tests: 62601
  Skipped          :    17 (0.03%)
  Unsupported      :  2518 (4.02%)
  Passed           : 59911 (95.70%)
  Expectedly Failed:   155 (0.25%)
[212/213] Running lit suite 

Testing Time: 12.47s

Total Discovered Tests: 8474
  Skipped:   17 (0.20%)
  Passed : 8457 (99.80%)
```

Extracted from #109032 (commit 3) (there are more refactors and cleanups in subsequent commits)

### CODE REVIEW

File: llvm/include/llvm/Transforms/Utils/Cloning.h [hunk 1]
```diff
@@ -207,6 +207,15 @@ void CloneAndPruneFunctionInto(Function *NewFunc, const Function *OldFunc,
                                const char *NameSuffix = "",
                                ClonedCodeInfo *CodeInfo = nullptr);
 
+/// Process function's subprogram attachment to collect relevant debug
+/// information in DIFinder.
```
Comment (felipepiovezan) [inline llvm/include/llvm/Transforms/Utils/Cloning.h:211]: Could you elaborate on "relevant debug info"? Reading this comment, I can't quite tell what the function is doing

File: llvm/lib/Transforms/Utils/CloneFunction.cpp [hunk 2]
```diff
@@ -137,6 +137,29 @@ void llvm::CloneFunctionAttributesInto(Function *NewFunc,
...
+  if (SPClonedWithinModule)
+    DIFinder.processSubprogram(SPClonedWithinModule);
+
+  const Module *M = F.getParent();
+  if (Changes != CloneFunctionChangeType::ClonedModule && M) {
+    // Inspect instructions to process e.g. DILexicalBlocks of inlined functions
+    for (const auto &BB : F) {
+      for (const auto &I : BB) {
+        DIFinder.processInstruction(*M, I);
+      }
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/Utils/CloneFunction.cpp:156]:
```
for (const auto &I : instructions(F))
  DIFinder...
```

[No need to go over BBs](https://llvm.org/doxygen/InstIterator_8h.html), also [no braces in single-line blocks](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements)

**Comment by felipepiovezan** [comment]
```
Test Plan:
ninja check-llvm-unit
```

Could you run the `check-llvm`? I don't think the unit tests will run the LIT tests as well, which impact the coro-splitter

File: llvm/lib/Transforms/Utils/CloneFunction.cpp (hunk 2 above)
Comment (artempyanykh) [reply llvm/lib/Transforms/Utils/CloneFunction.cpp]: Thanks! Updated.

**Comment by artempyanykh** [comment]
> Could you run `check-llvm`? I don't think the unit tests will run the LIT tests as well, which impact the coro-splitter

Sure! Did that and updated the test plan.

**Comment by artempyanykh** [comment]
Rebased on trunk

File: llvm/include/llvm/Transforms/Utils/Cloning.h [hunk 3]
```diff
@@ -207,6 +207,20 @@ void CloneAndPruneFunctionInto(Function *NewFunc, const Function *OldFunc,
                                const char *NameSuffix = "",
                                ClonedCodeInfo *CodeInfo = nullptr);
 
+/// Collect debug information such as types, compile units, and other
+/// subprograms that are reachable from \p F and can be considered global for
+/// the purposes of cloning (and hence not needing to be cloned).
+///
+/// The latter depends on \p Changes: when cloning into the same module we
```
Comment (felipepiovezan) [inline llvm/include/llvm/Transforms/Utils/Cloning.h:214]: Nit: it's unusual to use "the latter" after switching paragraphs. I'm assuming you are referring to "subprograms" here (note how far it is in the text), in which case we should replace "the latter" with "cloning subprograms"

**Review by felipepiovezan — APPROVED** [review]
LGTM but left one comment about documentation!

File: llvm/include/llvm/Transforms/Utils/Cloning.h (hunk 3 above)
Comment (artempyanykh) [reply llvm/include/llvm/Transforms/Utils/Cloning.h:214]: Good point @felipepiovezan! I'll fix that.

**Pull request merged**